src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -458,33 +464,46 @@ internal bool LocalFindAndPop(IThreadPoolWorkItem callback)
 
         public IThreadPoolWorkItem Dequeue(ThreadPoolWorkQueueThreadLocals tl, ref bool missedSteal)
         {
-            WorkStealingQueue localWsq = tl.workStealingQueue;
             IThreadPoolWorkItem callback;
-
-            if ((callback = localWsq.LocalPop()) == null && // first try the local queue
-                !workItems.TryDequeue(out callback)) // then try the global queue
+            int wsqActiveObserved = ThreadPoolGlobals.wsqActive;
+            if(wsqActiveObserved > 0)
             {
-                // finally try to steal from another thread's local queue
-                WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
-                int c = queues.Length;
-                Debug.Assert(c > 0, "There must at least be a queue for this thread.");
-                int maxIndex = c - 1;
-                int i = tl.random.Next(c);
-                while (c > 0)
+                WorkStealingQueue localWsq = tl.workStealingQueue;
+
+                if ((callback = localWsq.LocalPop()) == null && // first try the local queue
+                    !workItems.TryDequeue(out callback)) // then try the global queue
                 {
-                    i = (i < maxIndex) ? i + 1 : 0;
-                    WorkStealingQueue otherQueue = queues[i];
-                    if (otherQueue != localWsq && otherQueue.CanSteal)
+                    // finally try to steal from another thread's local queue
+                    WorkStealingQueue[] queues = WorkStealingQueueList.Queues;
+                    int c = queues.Length;
+                    Debug.Assert(c > 0, "There must at least be a queue for this thread.");
+                    int maxIndex = c - 1;
+                    int i = tl.random.Next(c);
+                    while (c > 0)
                     {
-                        callback = otherQueue.TrySteal(ref missedSteal);
-                        if (callback != null)
+                        i = (i < maxIndex) ? i + 1 : 0;
+                        WorkStealingQueue otherQueue = queues[i];
+                        if (otherQueue != localWsq && otherQueue.CanSteal)
                         {
-                            break;
+                            callback = otherQueue.TrySteal(ref missedSteal);
+                            if (callback != null)
+                            {
+                                break;
+                            }
                         }
+                        c--;
+                    }
+                    if((callback == null) && !missedSteal)
+                    {
+                        Interlocked.CompareExchange(ref ThreadPoolGlobals.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
```

Comment:
> LocalPush could just do a volatile write

It is an ECMA IL requirement that a volatile write constrains ordering and has Release semantics.  This is functioanlly correct on ARM64 and is required on all platforms.  I think the volatile write is sufficient.  However, I am still tempted to add `Volatile.Read(ThreadPoolGlobals.wsqActive)` in `Enqueue()` to be explicit and to prevent painful future debugging if something changes.

> isn't it possible that any thread that decrements wsqActive from 1 to 0 could miss a concurrently queued work item in a local queue that it already scanned before, and it would still decrement wsqActive to 0?

Arm loosely ordered memory semantics are most likely to be the problem and the hardest to analyse.

**From ARMv8 ARM**
```
Barrier instructions order prior Memory effects before subsequent Memory effects generated by the
same Observer.
```
```
The Memory effects of an instruction are the read, write, or barrier effects of that instruction. For an
instruction that accesses memory:
• A read effect is generated for each Location that is read by the instruction.
• A write effect is generated for each Location that is written by the instruction
```

**Design requirements**
For `Enqueue()` 
+ The write effect `LocalPush()` must be ordered before the read effect `Volatile.Read(ThreadPoolGlobals.wsqActive)`

For `Dequeue()` 
+ The write effect of the atomic decrement of `ThreadPoolGlobals.wsqActive` must be ordered before the read effects of `LocalPop()` & `CanSteal`

So if these are ordered as required by the design, these are the ordered operations in each thread:
 A. Enqueue() LocalPush()
 B. Enqueue() Volatile.Read(ThreadPoolGlobals.wsqActive) == 2

1. Dequeue() observe wsqActive=2
2. queue apparently empty, atomic decrement, wsqActive=1
3. Dequeue() observe wsqActive=1
4. queue apparently empty, atomic decrement, wsqActive=0

So for this race to occur
+ A must be ordered between 3 & 4. 
+ B must be ordered before 2

Per design, since B is ordered after A, the race cannot occur.

I need to make sure ordering design is implemented is correctly for Arm/Arm64.

When Arm/Arm64 is correct we may need to review AMD64.  My guess is the Arm64 solution will work for AMD64/x86